kinematic_equations returns zero when speeds is zero

arihantparsoya opened the pull request:
Fixes #12600. 

### Outpur before this PR:
``` python
>>> import sympy as sm
>>> import sympy.physics.mechanics as me
>>> N = me.ReferenceFrame('N')
>>> me.kinematic_equations((0, 0, 0), (sm.pi, sm.pi / 2, 0), 'body', '132')
[nan, 0, nan]
>>> A = N.orientnew('A', 'body', (sm.pi, sm.pi / 2, 0), '132')
>>> A.ang_vel_in(N)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/parsoyaarihant/sympy/sympy/physics/vector/vector.py", line 361, in __str__
    if ar[i][0][j] == 1:
TypeError: 'NaN' object does not support indexing
```

### Output after this PR:
``` python
>>> import sympy as sm
>>> import sympy.physics.mechanics as me
>>> N = me.ReferenceFrame('N')
>>> me.kinematic_equations((0, 0, 0), (sm.pi, sm.pi / 2, 0), 'body', '132')
[0, 0, 0]
>>> A = N.orientnew('A', 'body', (sm.pi, sm.pi / 2, 0), '132')
>>> A.ang_vel_in(N)
0
```

* Tests are added.

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

arihantparsoya committed:
`kinematic_equation()` returns zero when `speeds` is zero

Fixes #12600 .

Function `kinematic_equation()` should return zero if the input
`speeds` is zero regardless of any mathematical rules.

Added conditional in line 277 to check if `speeds` is zero. If it is,
then return [0,0,0] directly.

Tests are added.

arihantparsoya committed:
ang_vel_in() return zero for non-time varying orientation

As discussed in issue #12600. `ang_vel_in()` should return zero for
non-time-varying orientation.

I have fixed the issue in `orient()` instead of `kinematic_equations()`
since the function `orient()` uses `solve` method in line 630. When
`templist` is `[0,0,0]`, `solve` returns `[]` and cause error in lines
631-632.

Tests are added.

parsoyaarihant commented:
@moorepants , can you review this?

smichr commented on sympy/physics/vector/frame.py:
```diff
@@ -621,13 +621,16 @@ def _rot(axis, angle):
-                u1 = expand(td[u1])
-                u2 = expand(td[u2])
-                u3 = expand(td[u3])
+                if all(i.is_number for i in [q1, q2, q3]):
```
Should all `u` be zero if all `q` were 1?

parsoyaarihant replied:
@moorepants , suggested that:

> The automatic calculation of angular velocity relies on kinematic_equations. This should
> simply return a zero vector for any non-time varying orientation.

moorepants replied:
I'm not sure what is_number does. I think it should return zeros if all of the u's are not functions of dynamicsymbols._t.

smichr replied:
OK, then checking that they are numbers is the most restrictive case. Might a more lenient tests like `if not any(time_var in i.free_symbols for i in (q1, q2, q3)):` be better?

parsoyaarihant replied:
> if not any(time_var in i.free_symbols for i in (q1, q2, q3)):

@smichr , I was trying to do something like this but I was unable to figure out how to do it correctly. Apparently, `t` is a general SymPy symbol. So it is hard to distinguish it from other `free_symbols`.

smichr replied:
If it's a general symbol then it won't be clashing with any other symbol so that check should work. I guess the most definitive answer (since zeros can hide in t-dependent expressions) would be to use 
```
ui = list(symbols('u1, u2, u3', cls=Dummy))
for i, qi in enumerate((q1, q2, q3)):
    if qi.is_constant(t):
        ui[i] = 0
u1, u2, u3 = ui
```

(You'll have to define t as is done earlier in this section of code.)

smichr replied:
btw, if you could give #12683 or #12684 a look over, that would be appreciated.

smichr commented on sympy/physics/vector/frame.py:
```diff
@@ -621,13 +621,16 @@ def _rot(axis, angle):
                 from sympy.physics.vector.functions import kinematic_equations
                 q1, q2, q3 = amounts
                 u1, u2, u3 = symbols('u1, u2, u3', cls=Dummy)
```
nit: I would move this down to just after the `else` of the `if-else` block.

parsoyaarihant replied:
`u1, u2, u3` are being used in `else` condition hence they should be defined before `if-else` block.

smichr replied:
Yes. I meant put them in the `else` part of the block where they are used. In the `if` part they are being set to 0; they are used in the `else` part as you note.

smichr commented:
This is an alternate to #12612 and appears to be more general. The other issue should be closed if this is committed.

arihantparsoya committed:
checked if `q1, q2, q3` are independent of time

Based on discussion on github, I have declared time symbol `t`(line
624) and verified that `q1, q2, q3` are independent of `t`.

moorepants commented:
This looks like a good solution. I approve after @smichr's issue is addressed.

parsoyaarihant commented:
@moorepants , can you review the changes?

smichr commented:
Does it make sense to allow for the case that not all q are time independent? It looks like those that *do* depend on time need to be run through the solution process. You have handled the all-zero case. If any are not zero then the code as it is will treat them as *all* non-zero and we might just end up in the same place as we are now (i.e. running into a problem because we assumed that something was not zero when it really was.)  

I don't know a lot about this routine, so maybe there is a good reason to suspect an all or nothing sort of case (all) the time.

smichr commented:
BTW: If the code really shouldn't proceed unless it knows if q is t-dependent or not, then we should also consider the case when `foo.is_constant(t)` returns None and raise an Undecidable error.

moorepants commented:
I think these are the only cases:

- speeds are all expressions that include functions of time
- some speeds are expressions that include functions of time where as the remaining speeds are potentially expressions of things that aren't functions of time
- all speeds are not expressions that include functions of time

Only the last case should be treated special. If there is no dependence on time, then kinematic_equations should return all zeros.

smichr commented:
OK, I looked more closely at what is being done. `kinematic_equations` is trying to return the general solution for 3 equations. The problem is that if there are not 3 equations to be solved you are going to run into problems. Catching the case when there are no equations to be solved is 1/7th of the problem. There are 6 other cases wherein a problem might be encountered. Consider what happens if `q2` alone is independent of time and has a value of `pi/2`. That value is used to calculate `c2` *which is needed to calculate `q3d`* so we are still going to generate a division by 0.

```
        q1, q2, q3 = coords
        q1d, q2d, q3d = [diff(i, dynamicsymbols._t) for i in coords]
        w1, w2, w3 = speeds
        s1, s2, s3 = [sin(q1), sin(q2), sin(q3)]
        c1, c2, c3 = [cos(q1), cos(q2), cos(q3)]
        if rot_type.lower() == 'body':
            if rot_order == '123':
                return [q1d - (w1 * c3 - w2 * s3) / c2, q2d - w1 * s3 - w2 *
                        c3, q3d - (-w1 * c3 + w2 * s3) * s2 / c2 - w3]
```

If you are only going to fix 1/7th of the problem, I recommend putting the test for zeros in `kinematic_equations` right after the q1d values are computed and adding a note that other cases are not handled. I would recommend against that hack, however, and consider how the "right thing" can be done to fix this situation. (And I'm not yet sure that that is.)

smichr commented:
Oh, the fix is obvious: modify the equations so they don't involve division. e.g.

instead of this,
```
if rot_order == '123':
                return [q1d - (w1 * c3 - w2 * s3) / c2, q2d - w1 * s3 - w2 *
                        c3, q3d - (-w1 * c3 + w2 * s3) * s2 / c2 - w3]
```

this:

```
            if rot_order == '123':
                return [q1d*c2 - (w1 * c3 - w2 * s3), q2d - w1 * s3 - w2 *
                        c3, q3d*c2 - (-w1 * c3 + w2 * s3) * s2 - w3*c2]
```

moorepants commented:
I don't think that you can modify all of the equations to avoid division and singularities at certain angles. This expression that you modified is equated to [u1, u2, u3], so the multiplication by c2 has to be done on both sides of the equation. 

We also worked on the garbage in, garbage out principle to some degree, i.e. the user must choose a rotation order that is best suited for the motion of the system to avoid singularities. But a common need is to rotate reference frames wrt to each other through non-time dependent rotations. It is desirable for the vector module to be aware of reference frame relations of this nature and to properly set the angular velocity relations to reflect.

moorepants commented:
The kinematic_equations function could check for singularities (division by zero) and raise an error for any case other than the speeds all equal to zero.

smichr commented:
> This expression that you modified is equated to [u1, u2, u3], so the multiplication by c2 has to be 
> done on both sides of the equation.

I believe the equations *relate* `qdot` to `ui`. When solved, they define `ui`. Here is what happens to the list of equations stored in `templist`:
```
td = solve(templist, [u1, u2, u3])  
u1 = expand(td[u1])  
u2 = expand(td[u2])  
u3 = expand(td[u3])  
```
So the multiplying through that I did should be valid.

Let's look at the first case (where `wi` = `ui`):

```
if rot_order == '123': 
    return [q1d - (w1 * c3 - w2 * s3) / c2, q2d - w1 * s3 - w2 * 
                c3, q3d - (-w1 * c3 + w2 * s3) * s2 / c2 - w3] 
```
If `q1d = 0` then we will try to solve 
```
[ 0 - (w1 * c3 - w2 * s3) / c2, q2d - w1 * s3 - w2 * c3, q3d - (-w1 * c3 + w2 * s3) * s2 / c2 - w3]
``` 
for `w1`, `w2` and `w3`. We have 3 eqns in 3 unk so this should proceed fine unless `c2` also happens to be zero. 

If we had cleared the denominators of each expression at the start we would have solved
```
eqn1 = [c2*q1d - (w1 * c3 - w2 * s3), q2d - w1 * s3 - w2 * c3, q3d*c2 - (-w1 * c3 + w2 * s3) * s2 - w3*c2]
```
And then if `q1d=0`, `eqn1` becomes
```
[0 - (w1 * c3 - w2 * s3), q2d - w1 * s3 - w2 * c3, q3d*c2 - (-w1 * c3 + w2 * s3) * s2 - w3*c2]
```
and that has the solution
```
{w3:q3d, w1:q2d∗s3/(c3∗∗2+s3∗∗2), w2:c3∗q2d/(c3∗∗2+s3∗∗2)
```
Or, if c2 = 0, then `eqn1` becomes

```
[0 - (w1 * c3 - w2 * s3), q2d - w1 * s3 - w2 * c3, 0 - (-w1 * c3 + w2 * s3) * s2 - 0]
```
which is independent of w3, is only dependent on `q2d` and has the solution
```
{w1: q2d∗s3/(c3∗∗2+s3∗∗2), w2: c3∗q2d/(c3∗∗2+s3∗∗2)}
```
Does this physically mean that the solution does not depend on w3? If so, it could arbitrarily be set to 0 and the line that retrieves the code could be changed to 
```
u1 = expand(td.get(u1, 0))  
u2 = expand(td.get(u2, 0))  
u3 = expand(td.get(u3, 0))
```

smichr commented:
I am going to close in favor of #12698 which may eventually get finalized.
